subsys/bluetooth/host/l2cap.c

```diff
@@ -3119,57 +3114,43 @@ int bt_l2cap_chan_disconnect(struct bt_l2cap_chan *chan)
 static int bt_l2cap_dyn_chan_send(struct bt_l2cap_le_chan *le_chan, struct net_buf *buf)
 {
 	struct l2cap_tx_meta_data *data;
-	void *old_user_data = l2cap_tx_meta_data(buf);
-	uint16_t sdu_len;
-	int err;
 
-	sdu_len = net_buf_frags_len(buf);
+	LOG_DBG("chan %p buf %p", le_chan, buf);
 
-	if (sdu_len > le_chan->tx.mtu) {
+	if (net_buf_frags_len(buf) > le_chan->tx.mtu) {
+		LOG_ERR("attempt to send %u bytes on %u MTU chan",
+			net_buf_frags_len(buf), le_chan->tx.mtu);
 		return -EMSGSIZE;
 	}
 
+	if (net_buf_headroom(buf) < BT_L2CAP_SDU_CHAN_SEND_RESERVE) {
+		/* Call `net_buf_reserve(buf, BT_L2CAP_SDU_CHAN_SEND_RESERVE)`
+		 * before passing buffer to L2CAP.
+		 */
+		LOG_DBG("Not enough headroom in buf %p", buf);
+		return -EINVAL;
+	}
+
 	data = alloc_tx_meta_data();
 	if (!data) {
 		LOG_WRN("Unable to allocate TX context");
 		return -ENOBUFS;
 	}
 
-	/* Prepend SDU "header" */
-	net_buf_push_le16(buf, sdu_len);
+	/* Push the SDU header before putting `buf` on TX queue */
```

Comment:
I suggest removing this change.

Let's avoid saying what comes next, since it's happening right below. I.e. remove "before putting `buf` on TX queue". Comments have the potential to become stale or require bigger diffs, so they have to contribute more than they cost.

I suggest avoiding the word "push"; it's ambiguous. In this case it's clear, since the reader knows the headers are at the beginning, but..